File: Frontend/apps/astrogame_app/lib/models/resources/stellar_object_resource.dart
import 'package:astrogame_app/converters/guid_converter.dart';
import 'package:astrogame_app/models/resources/resource.dart';
import 'package:astrogame_app/models/stellar/base_types/stellar_object.dart';
import 'package:flutter_guid/flutter_guid.dart';
import 'package:json_annotation/json_annotation.dart';

part 'stellar_object_resource.g.dart';

@GuidConverter()
@JsonSerializable()
class StellarObjectResource {
  @JsonKey()
  Guid id;

  @JsonKey()
  Guid stellarObjectId;

  @JsonKey()
  Guid resourceId;

  @JsonKey()
  double multiplier;

  @JsonKey()
  StellarObject stellarObject;

  @JsonKey()
  Resource resource;

  StellarObjectResource(
    this.id,
    this.stellarObjectId,
    this.resourceId,
    this.multiplier,
    this.stellarObject,
    this.resource,
  );

  factory StellarObjectResource.fromJson(Map<String, dynamic> json) =>
      _$StellarObjectResourceFromJson(json);
  Map<String, dynamic> toJson() => _$StellarObjectResourceToJson(this);
}

File: Backend/AstroGame.Shared/Models/Resources/StellarObjectResource.cs
using AstroGame.Shared.Models.Stellar.BaseTypes;
using Newtonsoft.Json;
using System;

namespace AstroGame.Shared.Models.Resources
{
    /// <summary>
    /// The connection between a <see cref="StellarObject"/> and a <see cref="Resource"/>
    /// </summary>
    public class StellarObjectResource
    {
        public Guid Id { get; set; }

        public Guid StellarObjectId { get; set; }

        public Guid ResourceId { get; set; }

        public double Multiplier { get; set; }

        [JsonIgnore] public virtual StellarObject StellarObject { get; set; }

        public virtual Resource Resource { get; set; }
    }
}